File: solutions/ruby/bob/1/bob.rb
class Bob

  def self.hey(remark)
    if isSilence?(remark)
      "Fine. Be that way!"
    elsif isForcefullyQuestion?(remark)
      "Calm down, I know what I'm doing!"
    elsif isYelling?(remark)
      "Whoa, chill out!"
    elsif isQuestion?(remark)
      "Sure."
    else
      "Whatever."
    end 
  end

  private_class_method def self.isSilence?(remark)
    remark.strip == ""
  end

  private_class_method def self.isYelling?(remark)
    remark.match?(/[a-zA-Z]/) and remark == remark.upcase
  end

  private_class_method def self.isQuestion?(remark)
    remark.strip.end_with?("?")
  end

  private_class_method def self.isForcefullyQuestion?(remark)
    isQuestion?(remark) and isYelling?(remark)
  end
end

File: solutions/clojure/bob/1/src/bob.clj
(ns bob)
(use 'clojure.string)

(defn is-silence?  [s] (= "" (trim s)))
(defn is-question? [s] (ends-with? (trim s) "?"))
(defn is-shouting? [s] 
  (and 
    (= s (clojure.string/upper-case s))
    (not (= s (clojure.string/lower-case s)))
  )
)

(defn response-for [s]
  (if (is-silence? s)
    "Fine. Be that way!"
    (if (and (is-question? s) (is-shouting? s))
      "Calm down, I know what I'm doing!"
      (if (is-question? s)
        "Sure."
        (if (is-shouting? s) 
          "Whoa, chill out!"
          "Whatever."
        )
      )
    )
  )
)

; 